File: nn/tests/test_nn.py
"""tests/test_nn.py - Module, Linear, Embedding, RMSNorm, Adam tests."""
import numpy as np
from nn import Tensor, Module, Linear, Embedding, RMSNorm, Adam

def test_module_parameters():
    class DummyModule(Module):
        def __init__(self):
            self.p1 = Tensor([1.0], requires_grad=True)
            self.p2 = Tensor([2.0], requires_grad=True)
    
    m = DummyModule()
    params = m.parameters()
    assert len(params) == 2
    print("  [PASS] Module.parameters()")

def test_linear_no_bias():
    linear = Linear(3, 4)
    x = Tensor(np.random.randn(2, 3).astype(np.float32), requires_grad=True)
    y = linear(x)
    assert y.data.shape == (2, 4)
    params = linear.parameters()
    assert len(params) == 1
    print("  [PASS] Linear (no bias)")

def test_linear_with_bias():
    linear = Linear(3, 4, bias=True)
    x = Tensor(np.random.randn(2, 3).astype(np.float32), requires_grad=True)
    y = linear(x)
    assert y.data.shape == (2, 4)
    params = linear.parameters()
    assert len(params) == 2
    print("  [PASS] Linear (with bias)")

def test_embedding():
    embed = Embedding(10, 4)
    indices = Tensor([1, 3, 5, 3, 1], requires_grad=False)
    out = embed(indices)
    assert out.data.shape == (5, 4)
    print("  [PASS] Embedding")

def test_embedding_backward():
    embed = Embedding(10, 4)
    indices = Tensor([1, 3, 5], requires_grad=False)
    out = embed(indices)
    loss = out.sum()
    loss.backward()
    assert embed.weight.grad is not None
    print("  [PASS] Embedding backward")

def test_rmsnorm():
    norm = RMSNorm(4)
    x = Tensor(np.random.randn(2, 4).astype(np.float32), requires_grad=True)
    y = norm(x)
    assert y.data.shape == (2, 4)
    print("  [PASS] RMSNorm")

def test_rmsnorm_backward():
    norm = RMSNorm(4)
    x = Tensor([[1.0, 2.0, 3.0, 4.0]], requires_grad=True)
    y = norm(x)
    loss = y.sum()
    loss.backward()
    print("  [PASS] RMSNorm backward")

def test_adam_optimizer():
    p = Tensor([1.0, 2.0, 3.0], requires_grad=True)
    p.grad = np.array([0.1, 0.2, 0.3])
    optim = Adam([p], lr=0.01)
    old_data = p.data.copy()
    optim.step()
    assert not np.allclose(p.data, old_data)
    optim.zero_grad()
    assert np.allclose(p.grad, 0.0)
    print("  [PASS] Adam optimizer")

def test_linear_backward():
    linear = Linear(2, 1)
    x = Tensor([[1.0, 2.0]], requires_grad=True)
    y = linear(x)
    loss = y.sum()
    loss.backward()
    assert linear.weight.grad is not None
    print("  [PASS] Linear backward")

if __name__ == "__main__":
    print("\n=== nn Module Tests ===")
    test_module_parameters()
    test_linear_no_bias()
    test_linear_with_bias()
    test_embedding()
    test_embedding_backward()
    test_rmsnorm()
    test_rmsnorm_backward()
    test_adam_optimizer()
    test_linear_backward()
    print("\n  all passed")

File: nn/tests/test_nn.ts
import { Tensor, Module, Linear, Embedding, RMSNorm, Adam } from '../index';

function assert(condition: boolean, msg: string) {
  if (!condition) throw new Error(`Assertion failed: ${msg}`);
}

function testModuleParameters() {
  class DummyModule extends Module {
    p1: Tensor;
    p2: Tensor;
    constructor() {
      super();
      this.p1 = Tensor.from([[1]], true);
      this.p2 = Tensor.from([[2]], true);
    }
  }
  const m = new DummyModule();
  const params = m.parameters();
  assert(params.length === 2, `expected 2 params, got ${params.length}`);
  console.log('  [PASS] Module.parameters()');
}

function testLinearNoBias() {
  const linear = new Linear(3, 4);
  const x = Tensor.from([[1, 2, 3]], true);
  const y = linear.forward(x);
  assert(y.shape[0] === 1, `expected batch=1, got ${y.shape[0]}`);
  assert(y.shape[1] === 4, `expected out_features=4, got ${y.shape[1]}`);
  const params = linear.parameters();
  assert(params.length === 1, `expected 1 param, got ${params.length}`);
  console.log('  [PASS] Linear (no bias)');
}

function testLinearWithBias() {
  const linear = new Linear(3, 4, true);
  const x = Tensor.from([[1, 2, 3]], true);
  const y = linear.forward(x);
  assert(y.shape[0] === 1, `expected batch=1, got ${y.shape[0]}`);
  const params = linear.parameters();
  assert(params.length === 2, `expected 2 params, got ${params.length}`);
  console.log('  [PASS] Linear (with bias)');
}

function testEmbedding() {
  const embed = new Embedding(10, 4);
  const indices = Tensor.from([[1, 3, 5, 3, 1]]);
  const out = embed.forward(indices);
  assert(out.shape[0] === 5, `expected seq_len=5, got ${out.shape[0]}`);
  assert(out.shape[1] === 4, `expected embedding_dim=4, got ${out.shape[1]}`);
  console.log('  [PASS] Embedding');
}

function testEmbeddingBackward() {
  const embed = new Embedding(10, 4);
  const indices = Tensor.from([[1, 3, 5]]);
  const out = embed.forward(indices);
  const loss = out.sum();
  loss.backward();
  assert(embed.weight.grad !== undefined, 'expected weight.grad to be defined');
  console.log('  [PASS] Embedding backward');
}

function testRMSNorm() {
  const norm = new RMSNorm(4);
  const x = Tensor.from([[1, 2, 3, 4]], true);
  const y = norm.forward(x);
  assert(y.shape[0] === 1, `expected batch=1, got ${y.shape[0]}`);
  assert(y.shape[1] === 4, `expected dim=4, got ${y.shape[1]}`);
  console.log('  [PASS] RMSNorm');
}

function testRMSNormBackward() {
  const norm = new RMSNorm(4);
  const x = Tensor.from([[1, 2, 3, 4]], true);
  const y = norm.forward(x);
  const loss = y.sum();
  loss.backward();
  console.log('  [PASS] RMSNorm backward');
}

function testAdamOptimizer() {
  const p = Tensor.from([[1, 2, 3]], true);
  p.grad = [[0.1, 0.2, 0.3]];
  const optim = new Adam([p], 0.01);
  const oldData = p.data[0].slice();
  optim.step();
  const newData = p.data[0];
  const dataChanged = oldData.some((v, i) => v !== newData[i]);
  assert(dataChanged, 'expected data to change after step');
  optim.zeroGrad();
  const gradZero = p.grad[0].every(v => v === 0);
  assert(gradZero, 'expected grad to be zero after zeroGrad');
  console.log('  [PASS] Adam optimizer');
}

function testLinearBackward() {
  const linear = new Linear(2, 1);
  const x = Tensor.from([[1, 2]], true);
  const y = linear.forward(x);
  const loss = y.sum();
  loss.backward();
  assert(linear.weight.grad !== undefined, 'expected weight.grad to be defined');
  console.log('  [PASS] Linear backward');
}

console.log('\n=== nn Module Tests ===');
testModuleParameters();
testLinearNoBias();
testLinearWithBias();
testEmbedding();
testEmbeddingBackward();
testRMSNorm();
testRMSNormBackward();
testAdamOptimizer();
testLinearBackward();
console.log('\n  all passed');